File: app/src/main/java/pozzo/apps/travelweather/forecast/model/Weather.java
package pozzo.apps.travelweather.forecast.model;

import java.util.Arrays;
import java.util.List;

import pozzo.apps.travelweather.map.model.Address;

/**
 * This will represent the weather for a location in a period of time.
 */
public class Weather {
    public interface Col {
        String ADDRESS = "address";
        String FORECASTS = "forecasts";
        String URL = "url";
    }

    private Address address;
    private Forecast[] forecasts;
    private String url;

    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }

    public Forecast[] getForecasts() {
        return forecasts;
    }

    public void setForecasts(Forecast[] forecasts) {
        this.forecasts = forecasts;
    }

    public void setForecasts(List<Forecast> forecasts) {
        this.forecasts = forecasts.toArray(new Forecast[0]);
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Weather weather = (Weather) o;

        if (address != null ? !address.equals(weather.address) : weather.address != null)
            return false;
        // Probably incorrect - comparing Object[] arrays with Arrays.equals
        if (!Arrays.equals(forecasts, weather.forecasts)) return false;
        return !(url != null ? !url.equals(weather.url) : weather.url != null);

    }

    @Override
    public int hashCode() {
        int result = address != null ? address.hashCode() : 0;
        result = 31 * result + (forecasts != null ? Arrays.hashCode(forecasts) : 0);
        result = 31 * result + (url != null ? url.hashCode() : 0);
        return result;
    }
}

File: app/src/main/kotlin/pozzo/apps/travelweather/forecast/model/Weather.kt
package pozzo.apps.travelweather.forecast.model

import com.google.android.gms.maps.model.LatLng
import pozzo.apps.travelweather.map.model.Address
import java.util.*

data class Weather(
    val url: String,
    val forecasts: List<Forecast>,
    val address: Address,
    val poweredBy: PoweredBy) {

    companion object {
        private const val PAST_LIMIT = 2L * 60L * 60L * 1000L
    }

    val latLng: LatLng
        get() = address.latLng

    fun getForecast(date: Calendar): Forecast {
        var closer: Forecast? = null
        forecasts.forEach {
            val isInClosePastOrFuture = date.timeInMillis < it.dateTime.timeInMillis + PAST_LIMIT
            val isEvenCloser = it.dateTime.timeInMillis < closer?.dateTime?.timeInMillis ?: Long.MAX_VALUE
            if (isInClosePastOrFuture && isEvenCloser) {
                closer = it
            }
        }
        return closer ?: forecasts.last()
    }
}

//Passo 3 - E posso deixar o fato da progressao pela rota como um proximo passo?
//Mas uma forma bem simples seria pegar sempre o proximo...
//Passo 4+ - E num futuro ainda mais distante da pra tentar usar o google direction para uma precisao de tempo melhor
